onmt/translate/translator.py

```diff
@@ -1228,3 +1229,94 @@ def _score_target(
+    from sacrebleu import corpus_bleu
+
+    def __init__(self, fields, opt):
+        self.fields = fields
+        self.opt = opt
+        import pyonmttok
+        self.tgt_tokenizer = pyonmttok.Tokenizer(
+            **self.opt.tgt_onmttok_kwargs)
+
+    def tokenize_batch(self, batch_side, side):
+        field = dict(self.fields)[side].base_field
+        tokenized_sentences = []
+        for i in range(batch_side.shape[1]):
+            tokens = []
+            for t in range(batch_side.shape[0]):
+                token = field.vocab.itos[batch_side[t, i, 0]]
+                if token == field.pad_token or token == field.eos_token:
+                    break
+                if token != field.init_token:
+                    tokens.append(token)
+            tokenized_sentences.append(tokens)
+        return tokenized_sentences
+
+    def build_sources_and_refs(self, batch, mode):
+        if mode == 'valid':
+            sources = []
+            refs = []
+            for example in batch.dataset.examples:
+                sources.append(example.src[0])
+                refs.append(example.tgt[0])
+        elif mode == 'train':
+            sources = self.tokenize_batch(batch.src[0], 'src')
+            refs = self.tokenize_batch(batch.tgt, 'tgt')
+        return sources, refs
+
+    def translate(self, model, batch, gpu_rank, mode):
+        from onmt.translate.translator import Translator
+        model_opt = self.opt
+        from onmt.utils.parse import ArgumentParser
+        parser = ArgumentParser()
+        onmt.opts.translate_opts(parser)
+        base_args = (["-model", "dummy"] + ["-src", "dummy"])
+        opt = parser.parse_args(base_args)
+        opt.gpu = gpu_rank
+        ArgumentParser.validate_translate_opts(opt)
+        ArgumentParser.update_model_opts(model_opt)
+        ArgumentParser.validate_model_opts(model_opt)
+        scorer = onmt.translate.GNMTGlobalScorer.from_opt(opt)
+        import os
+        import codecs
+        out_file = codecs.open(os.devnull, "w", "utf-8")
+        translator = Translator.from_opt(
+            model,
+            self.fields,
+            opt,
+            model_opt,
+            global_scorer=scorer,
+            out_file=out_file,
+            report_align=opt.report_align,
+            report_score=True,
+            logger=None)
+        sources, refs = self.build_sources_and_refs(batch, mode)
+        _, preds = translator.translate(
+            sources,
+            batch_size=model_opt.valid_batch_size,
+            batch_type=model_opt.batch_type)
+        texts_ref = []
+
+        for i in range(len(preds)):
+            preds[i] = self.tgt_tokenizer.detokenize(preds[i][0].split())
+            texts_ref.append(self.tgt_tokenizer.detokenize(refs[i]))
+        try:
+            print("first source: ", sources[0])
+            print("first detok prediction: ", preds[0])
+            print("first ref: ", texts_ref[0])
+        except Exception:
+            pass
```

Comment:
This looks strange. Why should we expect errors there? And especially why should we let them pass instead of raising explicitly?

Reply:
Also, this should probably be logged with `logger`, and only in "debug" level ? Or even better, make a flag to allow to dump these predictions to facilitate inspection if needed.